database-commons/src/main/java/io/cdap/plugin/db/source/AbstractDBSource.java

```diff
@@ -262,11 +279,20 @@ private DriverCleanup loadPluginClassAndGetDriver(Class<? extends Driver> driver
     }
   }
 
-  private Connection getConnection() throws SQLException {
+  private Connection getConnection() {
     String connectionString = createConnectionString();
     Properties connectionProperties = new Properties();
     connectionProperties.putAll(sourceConfig.getConnectionArguments());
-    return DriverManager.getConnection(connectionString, connectionProperties);
+    return createConnectionWithRetry(connectionString, connectionProperties);
+  }
+
+  private Connection createConnectionWithRetry(String connectionString, Properties connectionProperties) {
+    return Failsafe.with(retryPolicy).<Connection>get(() -> DriverManager
+      .getConnection(connectionString, connectionProperties));
+  }
+
+  private Statement createStatementWithRetry(Connection connection) {
+    return Failsafe.with(retryPolicy).<Statement>get(() -> connection.createStatement());
```

Comment:
we should throw `cause` instead of `FailsafeException`.

Reply:
Handled the `FailsafeException` in a separate method : `unwrapFailsafeException` of `RetryUtils` class as suggested in the other comment.